lib/starttls.js

```diff
@@ -33,111 +15,46 @@ function starttlsClient(socket, options, callback) {
 }
 
 function starttls(socket, options, callback, isServer) {
-  var sslcontext;
-
   var opts = {};
+
   Object.keys(options).forEach(function(key) {
     opts[key] = options[key];
   });
   if (!opts.ciphers) {
     opts.ciphers = RECOMMENDED_CIPHERS;
   }
+  opts.isServer = isServer;
+  opts.secureContext = tls.createSecureContext(opts);
 
-  socket.removeAllListeners('data');
-  if (tls.createSecureContext) {
-    sslcontext = tls.createSecureContext(opts);
-  } else {
-    sslcontext = crypto.createCredentials(opts);
-  }
-  var pair = tls.createSecurePair(sslcontext, isServer);
-  var cleartext = pipe(pair, socket);
-
+  var secureSocket = new tls.TLSSocket(socket, opts);
   var erroredOut = false;
-  pair.on('secure', function() {
+
+  // NodeJS documentation bug: secure vs secureConnect
+  // https://github.com/nodejs/node/issues/10555
+  secureSocket.on('secure', function() {
     if (erroredOut) {
-      pair.end();
+      secureSocket.end();
       return;
     }
 
-    var verifyError = (pair._ssl || pair.ssl).verifyError();
-
-    if (verifyError) {
-      cleartext.authorized = false;
-      cleartext.authorizationError = verifyError;
+    var authError = secureSocket.ssl.verifyError();
+    if (authError) {
+        secureSocket.authorized = false;
+        secureSocket.authorizationError = authError;
     } else {
-      cleartext.authorized = true;
+        secureSocket.authorized = true;
```

Comment:
I agree that it is not very clear what should be subject to verification. We should keep in mind that when TLS handshake happens on control connection we don't even know client's login credentials. I'm sure it will be much easier to add this feature when there is a need for it, rather than if one has to guess what might be needed without any particular use case. I have updated my PR to include documentation changes clarifying how tlsOptions should be used and inserted a couple of TODOs to source code - hints on possible future improvements. Thanks!